client/client.go

```diff
@@ -213,10 +208,123 @@ func (client *Client) RemoveToxic(options *ToxicOptions) error {
 
 // ResetState resets the state of all proxies and toxics in Toxiproxy.
 func (client *Client) ResetState() error {
-	resp, err := http.Post(client.endpoint+"/reset", "text/plain", bytes.NewReader([]byte{}))
+	_, err := client.post("/reset", bytes.NewReader([]byte{}))
+	return err
+}
+
+func (c *Client) get(path string) ([]byte, error) {
+	req, err := http.NewRequest("GET", c.endpoint+path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.send(req, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	result, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (c *Client) post(path string, body io.Reader) ([]byte, error) {
+	req, err := http.NewRequest("POST", c.endpoint+path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.send(req, http.StatusCreated)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	result, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (c *Client) patch(path string, body io.Reader) ([]byte, error) {
+	req, err := http.NewRequest("PATCH", c.endpoint+path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.send(req, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	result, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func (c *Client) delete(path string) error {
+	req, err := http.NewRequest("DELETE", c.endpoint+path, nil)
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.send(req, http.StatusNoContent)
 	if err != nil {
 		return err
 	}
 
-	return checkError(resp, http.StatusNoContent, "ResetState")
+	resp.Body.Close()
+
+	return nil
+}
+
+func (c *Client) send(req *http.Request, success_code int) (*http.Response, error) {
+	req.Header.Set("User-Agent", c.UserAgent)
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := c.http.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("fail to request: %w", err)
+	}
+
+	err = c.validateResponse(resp, success_code)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+func (c *Client) validateResponse(resp *http.Response, expectedCode int) error {
+	if resp.StatusCode < 300 {
```

Comment:
to indicate success, I think `if resp.StatusCode >= http.StatusOK && resp.StatusCode < http.Status StatusMultipleChoices` would be more precise, to exclude `1xx`, not sure if that edge cases would ever happen

Reply:
I think it is too much. Toxiproxy does not have any 1xx response codes.
Updated the validation though.

I would still use numbers 200 and 300, instead of constants - just for simple readability.